### PR TITLE
introduce unrecoverable error type, and logs to show Fluentd process/worker started

### PR DESCRIPTION
**Pull request by tagomoris**
Currently, unrecoverable errors (which serverengine doesn't try to restart workers about) are only configuration errors.
But some other situation cannot be retried. For example:
* per-worker root directory (will be introduced next) exists but it is a file
* failed to bind an address/port in server plugin helper
* failed to read a file from in_tail by permission error
* or many others

We are using ConfigError for such cases, but it's nonsense in fact.

This change also adds unrecoverable cases for scripting errors while loading plugin files.
Fluentd is now restarting for such cases, but it's just wrong behavior.

### CODE REVIEW

**Comment by tagomoris** [comment]
Actually, the biggest improvement in this change is testing about:
* Fluentd process started correctly
* Fluentd process failed to start correctly

**Comment by tagomoris** [comment]
@repeatedly could you review this change?

File: lib/fluent/supervisor.rb [hunk 1]
```diff
@@ -617,39 +618,54 @@ def flush_buffer
...
+
     def main_process(&block)
       Process.setproctitle("worker:#{@process_name}") if @process_name
 
-      configuration_error = false
+      unrecoverable_error = false
 
       begin
         block.call
       rescue Fluent::ConfigError
```
Comment (repeatedly) [inline lib/fluent/supervisor.rb]: Use `Fluent::ConfigError => e` and remove `$!`.

File: lib/fluent/supervisor.rb [hunk 2]
```diff
@@ -617,39 +618,54 @@ def flush_buffer
...
-          log.level = @log_level
-          console = log.enable_debug
-          console.error "config error", file: @config_path, error: $!.to_s
-          console.debug_backtrace
+        logging_with_console_output do |log|
+          log.error "config error", file: @config_path, error: $!
+          log.debug_backtrace
+        end
+        unrecoverable_error = true
+      rescue Fluent::UnrecoverableError
```
Comment (repeatedly) [inline lib/fluent/supervisor.rb]: ditto

File: lib/fluent/supervisor.rb [hunk 3]
```diff
@@ -617,39 +618,54 @@ def flush_buffer
...
+        logging_with_console_output do |log|
+          log.error $!.message, error: $!
+          log.error_backtrace
+        end
+        unrecoverable_error = true
+      rescue ScriptError # LoadError, NotImplementedError, SyntaxError
```
Comment (repeatedly) [inline lib/fluent/supervisor.rb]: ditto

File: test/command/test_fluentd.rb [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+        "error_class=SyntaxError",
+        "in_buggy.rb:5: syntax error, unexpected end-of-input, expecting keyword_end",
+      )
+      assert !process_exist?(@pid)
+    end
+  end
+
+  sub_test_case 'configuration to load plugin which raises unrecoverable error in #start' do
+    test 'failed to start' do
+      script =  "require 'fluent/plugin/input'\n"
```
Comment (repeatedly) [inline test/command/test_fluentd.rb]: Use heredocument is better.

**Comment by repeatedly** [comment]
Others are LGTM and adding start test sis awesome.
BTW, test on Windows is unstable...

File: test/command/test_fluentd.rb (hunk 4 above)
Comment (tagomoris) [inline test/command/test_fluentd.rb]: Writing ruby code in here document breaks indentation on my environment... :P

**Comment by tagomoris** [comment]
I added some fixes about executing worker processes and showing process ids in log.
@repeatedly could you review these changes again?

**Comment by repeatedly** [comment]
LGTM and good fix for windows environment 👍

**Pull request merged**